proposals/xxxx-experimental-dxil.md

```diff
@@ -0,0 +1,195 @@
+seaks to implemented a systematic method to handle these issues.
+
+## Goals
+This proposal seeks to address the following points:
+ * Needless churn when experimental op are delayed or rejected
+ * Expermental op boundary point is rigid and moves with every SM update
+ * Long term experiments (and potentially extensions) aren't currently feasible
+ * Focused on core api system (hlsl instrinsics and DXIL ops)
+ * Works within the current intrinsics/DXIL op mechanisms
+ * Minimizes overall changes to the system and IHV processes
+ * Straightforward transition route from experimental to stable
+ * IHV drivers can support both experimental and stable versions of an op simultaneously
+   * simplifies migrations
+
+## Non-goals
+Future proposals may address the topics below but this proposal seeks to be a
+smaller isloated change. It intends to solves immediate term challenges
+without investing significant engineering efforts into a generalized solution.
+That said, an attempt is made to avoid proposals that preclude a generalized
+solution. Thus this proposal explicitly avoids addressing these issues:
+ * Full scale generalized extension system
+ * Process development to enable asynchronous non-colliding development
+ * Metadata/RDAT/PSV0/Custom lowering are out of scope for this document
+
+
+
+## Potential DXIL Op Solutions
+
+### Top 1 bit as "is experimental" flag
+
+The top bit of all opcodes is a flag stating if the opcode is experimental.
+
+No structural or shape changes to the DXIL occur, simply the fact that the opcode
+has the high bit set informs that it is experimental. This makes it very easy
+for the compiler and drivers to detect experimental opcodes. When an opcode is
+transistioned to stable the opcode needs to be assigned a stable number.
+This splits the 4 billion opcode space into two 2 billion partions. One for
+stable one for experimental.
+
+This is the simpliest proposal with the least invasive set of changes.
+
+Pros:
+ * Very simple
+ * Quick to implement
+ * Could be implemented "by hand" today by hard coding opcodes
+Cons:
+ * Not a solution for extensions
+ * transistion from experimental to stable isn't just unsetting the bit
+   * other stable ops may have already taken that number
+   * complicates the experimental->stable mapping
+
+### Top 8 bits as "opcode partition" value
+This is pretty much identical to the 1 bit flag proposal except there are 256
+partitions with 16 million opcodes each. The key difference is that it unlocks
+extension potential as extension developers such as IHVs could reserve a
+partition for their own use without collision with other opcodes.
+
+| Partition | Use |
+|-----------|-----|
+| 0 | stable |
+| 1 | experimental |
+| 2 | extension foo |
+| .. | extension .. |
+| 255 | extension 255 |
+
+
+Pros:
+ * Fairly simple
+ * Quick to implement
+ * Enables basic opcodes extension system
+Cons:
+ * transistion from experimental to stable isn't just clearing the partition
```

Comment:
As I mentioned before, I don't think there is any *solution* possible that could avoid this con.

Even then, "just clearing the partition" is changing the opcode, thus the IR, and *still* requires test updates to deal with. So, what's the difference between a hypothetical solution that avoids this con and one that has the con?

Are you imagining a hypothetical solution that I'm not?

Reply:
None of the "encode into the opcode" solutions avoid this con, but some of the more "robust" solutions actually can. Specifically, `### Single Specific Experimental Opcode with varargs` where there is never a "transition" period. All opcodes for the rest of time are "extension" opcodes and the lowering doesn't change. Then we maintain a list somewhere else that determines which extension sets are experimental and which are not and add infrastructure around that to ensure non-experimental sets aren't allowed in non-experimental builds

But as you said, that solution is a non-starter. Included this as a con was to compare against that and call out that a solution chosen with this con will need to resolve this issue in one manner or another

Reply:
Closing as has been discussed